### ISSUE TITLE
Symbol literals should have useful singleton types

### ISSUE THREAD

**Issue by milessabin** [typelevel/scala#51]
Scala `Symbols` have first class literal syntax like strings, however unlike string literals `Symbol` literals lack useful singleton types,

``` scala
scala> def single[T <: AnyRef](t: T): Option[t.type] = Some(t)
single: [T <: AnyRef](t: T)Option[t.type]

scala> single("foo")
res0: Option[String("foo")] = Some(foo)

scala> single('foo)
res1: Option[Symbol] = Some('foo)
```

shapeless simulates singleton types for `Symbol` literals by @@-tagging them with the singleton type of the corresponding string literal. It would be nice, particularly in the light of #15 and #27, to have richer native singleton types for `Symbol` literals too.

**Comment by propensive** [comment]
I would be very interested in seeing more general support for returning singleton types from user-defined StringContext "literals". This should be possible already using whitebox macros, though if that could be done avoiding all the macro machinery, that would be much preferred.

e.g.

```
scala> single(uri"http://www.example.com/")
res0: Option[HttpUrl("http://www.example.com/")] = Some(http://www.example.com/)
```

**Comment by milessabin** [comment]
This is fixed in the SIP-23 PR which will be included in the 2.12.0-RC1 Typelevel Scala release.
